jupyterhub/app.py

```diff
@@ -2128,6 +2128,7 @@ async def init_role_assignment(self):
                             if kind == 'users':
                                 orm_obj = await self._get_or_create_user(bname)
                                 orm_role_bearers.append(orm_obj)
+                                roles.grant_role(db, orm_obj, 'user')
```

Comment:
I think this makes sense, but can you move this assignment to within `_get_or_create_user` and only `if not username in self.authenticator.allowed_users:`?

Reply:
> move this assignment to within _get_or_create_user

In some cases `_get_or_create_user` is called before default roles (ex. `'user'`) are created. So I cannot move it.


> only if not username in self.authenticator.allowed_users:

You seem to be worried that `role.grant_role` is called twice. `role.grant_role` is idempotent. I think that it is better to let the code simple.

Reply:
> In some cases _get_or_create_user is called before default roles (ex. 'user') are created.

That [shouldn't be the case](https://github.com/jupyterhub/jupyterhub/blob/95649a3ece8532aa6d3e6bb47973434930fd61b8/jupyterhub/app.py#L2763-L2765), but it might be a missing call to `init_role_creation()` in some tests that use only a partially initialized Hub.

> You seem to be worried that role.grant_role is called twice

No, I'm just trying to get the code in the right place. This fixes one instance of the bug, but the same bug occurs in two places (here and in load_groups) - both places call `_get_or_create_user`, so that seems like the logical place - it shouldn't matter _why_ a User is getting created. Plus, this code involved db queries on a per-user basis, so there's a cost to calling it unnecessarily, even if it is idempotent.

Reply:
> That shouldn't be the case, but it might be a missing call to init_role_creation() in some tests that use only a partially initialized Hub.

I was too careful because of the test case. I don't have enough understanding of this project to modify existing test cases. I agree with your idea now.